File: 041 First Missing Positive.c
#include <stdio.h>
#include <stdlib.h>
/*
First Missing Positive
*/
#define DEBUG 1

void swap(int* i,int* j){
    int t = *i;
    *i    = *j;
    *j    = t;
}
void dump(int* a, int N)
{
    int i = 0;
    for(;i<N;i++) printf("%d ",a[i]);
    printf("\n");
}
int firstMissingPositive(int* arr, int size)
{
    int i,j,k;
    
    // 1. Move all negative numbers to right
    i = 0;
    j = size-1;
    while(i<=j){
        while(i<=j && arr[i] >= 0) i++;
        while(i<=j && arr[j] <  0) j--;
        if(i<j) swap(arr+i,arr+j);
    }
    
    // All negative or 0, $i never moved, $j is at front
    if(i == 0 && j == -1) return 1; 
    if(DEBUG){
        printf("First: ");
        dump(arr,size);
        printf("i: %d, j: %d\n",i,j);
    }
        
    // 2. Assign arr[k-1] to negative if visited
    for(i = 0;i<=j;i++){
		k = abs(arr[i]);
        if(k <= size && k > 0) // ignore k == 0
        {
            if(arr[k-1] > 0) arr[k-1] = -arr[k-1];
            else if(arr[k-1] == 0) arr[k-1] = -(size+1);
        }
		
		if(DEBUG){
		    printf("k : %d\n",k);
		    dump(arr,size);
		}
    }
    
    // 3. Find first non-negative
    for(i = 0;i<=j;i++)
        if(arr[i]>=0)
            return i+1;
    
    return j+2;
}
/*
Tester
*/
int main()
{
    //int arr[] = {0,5,-2,3,7,1,2,-15};
    //int arr[] = {0,-10,1,-5,0,2,3,4};
    //int arr[] = {1,2,0}; // 3
    //int arr[] = {1}; // 2
    //int arr[] = {1,2,3}; // 4
    //int arr[] = {0}; // 1
    //int arr[] = {2,2}; // 1
    //int arr[] = {-10,-3,-100,-1000,-239,1};
    int arr[] = {1,0,3,3,0,2};
    int arr_size = sizeof(arr)/sizeof(arr[0]);
    printf("%d\n", firstMissingPositive(arr, arr_size));
    return 0;
}

File: 041 First Missing Positive.py
'''
41. First Missing Positive
https://leetcode.com/problems/first-missing-positive/

Given an unsorted integer array nums, return the smallest missing positive integer.

You must implement an algorithm that runs in O(n) time and uses O(1) auxiliary space.

Example 1:

Input: nums = [1,2,0]
Output: 3
Explanation: The numbers in the range [1,2] are all in the array.

Example 2:

Input: nums = [3,4,-1,1]
Output: 2
Explanation: 1 is in the array but 2 is missing.

Example 3:

Input: nums = [7,8,9,11,12]
Output: 1
Explanation: The smallest positive integer 1 is missing.

Constraints:
    1 <= nums.length <= 105
    -231 <= nums[i] <= 231 - 1
'''
from typing import List

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            # Swap a positive into its right place, for example, 3 should be put into index 2
            # Do this in while loop, because nums[nums[i]-1] might need to be swapped too.
            # This runs in O(n) as each number is swapped at most once
            while nums[i] > 0 and nums[i] <= len(nums) and nums[nums[i]-1] != nums[i]:
                nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1]
            
        # Check from beginning
        for i in range(len(nums)):
            if nums[i] != i + 1:
                return i + 1
        else:
            return len(nums) + 1

   
'''
Real facebook question:
第二题 输入是一个数组和数字n, 如果数组里正好有1到n的整数输出true否则false
https://www.1point3acres.com/bbs/thread-1041414-1-1.html
'''
def checkN(arr: List[int], n: int) -> bool:
    if len(arr) != n:
        # 1. Not same length
        return False
    
    for i in range(n):
        # Find the right element for index i
        while arr[i] != i + 1:
            if not 1 <= arr[i] <= n:
                # 2. The element is out of bound
                return False
        
            right = arr[i] - 1
            if arr[right] == arr[i]:
                # 3. Duplicates
                return False
            
            arr[right], arr[i] = arr[i], arr[right]

    return True

cases = [
    [1,2,3,4,5],
    [5,4,3,2,1],
    [1,1,1,3,2],
    [5,4,1,3,6],
    [5,4,2,3,7],
    [4,3,1,5,2],
]
for case in cases:
    # Interesting enough, Python executes the part inside 'format' first, so
    # the input array is changed
    print('input: {0}, output: {1}'.format(case, checkN(case, len(case))))